text/0000-trim-std-os.md

```diff
@@ -0,0 +1,146 @@
+  use LFS functions][std-no-lfs], so files over 4GB cannot be opened. Changing
+  this, however, would [involve changing the `stat`
+  structure][libc-stat-change] and may be difficult to do.
+* Trait extensions in the `raw` module attempt to return the `libc` aliased type
+  on all platforms, for example [`DirEntryExt::ino`][std-nio] returns a type of
+  `ino_t`.  The `ino_t` type is billed as being FFI compatible with the libc
+  `ino_t` type, but not all platforms store the `d_ino` field in `dirent` with
+  the `ino_t` type. For example on Android the [definition of
+  `ino_t`][android-ino_t] is `u32` but the [actual stored value is
+  `u64`][android-d_ino]. This means that on Android we're actually silently
+  truncating the return value!
+
+[lfs]: http://users.suse.com/~aj/linux_lfs.html
+[std-as-stat]: https://github.com/rust-lang/rust/blob/29ea4eef9fa6e36f40bc1f31eb1e56bf5941ee72/src/libstd/sys/unix/fs.rs#L81-L92
+[std-no-lfs]: https://github.com/rust-lang/rust/issues/30050
+[std-ino]: https://github.com/rust-lang/rust/blob/29ea4eef9fa6e36f40bc1f31eb1e56bf5941ee72/src/libstd/sys/unix/fs.rs#L192-L197
+[libc-stat-change]: https://github.com/rust-lang-nursery/libc/blob/2c7e08c959e599ca221581b1670a9ecbbeac2dcb/src/unix/notbsd/linux/other/b32/mod.rs#L28-L71
+[android-d_ino]: https://github.com/rust-lang-nursery/libc/blob/2c7e08c959e599ca221581b1670a9ecbbeac2dcb/src/unix/notbsd/android/mod.rs#L50
+[android-ino_t]: https://github.com/rust-lang-nursery/libc/blob/2c7e08c959e599ca221581b1670a9ecbbeac2dcb/src/unix/notbsd/android/mod.rs#L11
+
+Over time it's basically turned out that exporting the somewhat-messy details of
+libc has gotten a little messy in the standard library as well. Exporting this
+functionality (e.g. being able to access all of the fields), is quite useful
+however! This RFC proposes tweaking the design of the extensions in
+`std::os::*::raw` to allow the same level of information exposure that happens
+today but also cut some of the tie from libc to std to give us more freedom to
+change these implementation details and work around weird platforms.
+
+# Detailed design
+[design]: #detailed-design
+
+First, the types and type aliases in `std::os::*::raw` will all be
+deprecated. For example `stat`, `ino_t`, `dev_t`, `mode_t`, etc, will all be
+deprecated (in favor of their definitions in the `libc` crate). Note that the C
+integer types, `c_int` and friends, will not be deprecated.
+
+Next, all existing extension traits will cease to return platform specific type
+aliases (such as the `DirEntryExt::ino` function). Instead they will return
+`u64` across the board unless it's 100% known for sure that fewer bits will
+suffice. This will improve consistency across platforms as well as avoid
+truncation problems such as those Android is experiencing. Furthermore this
+frees std from dealing with any odd FFI compatibility issues, punting that to
+the libc crate itself it the values are handed back into C.
```

Comment:
@tbu- yeah the burden would be on consumers of libstd to cast appropriately and panic if out-of-bounds. I would expect that 95% of the time the types would always fit, but for the cases like LFS that @cuviper mentions it could be that the two applications are compiled differently (e.g. linking to C which isn't using LFS and hence can't handle file sizes > 4GB on 32-bit Linux).